pkg/modifycontroller/controller.go

```diff
@@ -0,0 +1,334 @@
+	eventRecorder := eventBroadcaster.NewRecorder(scheme.Scheme,
+		v1.EventSource{Component: fmt.Sprintf("external-resizer %s", name)})
+
+	claimQueue := workqueue.NewNamedRateLimitingQueue(
+		pvcRateLimiter, fmt.Sprintf("%s-pvc", name))
+	vacQueue := workqueue.NewNamedRateLimitingQueue(
+		pvcRateLimiter, fmt.Sprintf("%s-vac", name))
+
+	ctrl := &modifyController{
+		name:        name,
+		modifier:    modifier,
+		kubeClient:  kubeClient,
+		pvInformer:  pvInformer.Informer(),
+		pvcInformer: pvcInformer.Informer(),
+		vacInformer: vacInformer.Informer(),
+		// pvSynced:                pvInformer.Informer().HasSynced,
+		claimQueue: claimQueue,
+		vacQueue:   vacQueue,
+		// volumes:                 pvInformer.Informer().GetStore(),
+		eventRecorder: eventRecorder,
+	}
+
+	// Add a resync period as the PVC's request modify can be modified again when we handling
+	// a previous modify request of the same PVC.
+	pvcInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ctrl.addPVC,
+		UpdateFunc: ctrl.updatePVC,
+		DeleteFunc: ctrl.deletePVC,
+	}, resyncPeriod)
+
+	// Add a resync period as the VAC can be created after a PVC is created
+	// However since VAC is immutable, there should have no update
+	vacInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ctrl.addVAC,
+		DeleteFunc: ctrl.deleteVAC,
+	}, resyncPeriod)
+
+	return ctrl
+}
+
+func (ctrl *modifyController) initUncertainPVCs() error {
+	ctrl.uncertainPVCs = make(map[string]v1.PersistentVolumeClaim)
+	for _, pvcObj := range ctrl.pvcInformer.GetStore().List() {
+		pvc := pvcObj.(*v1.PersistentVolumeClaim)
+		if pvc.Status.ModifyVolumeStatus != nil && (pvc.Status.ModifyVolumeStatus.Status == v1.PersistentVolumeClaimModifyVolumeInProgress || pvc.Status.ModifyVolumeStatus.Status == v1.PersistentVolumeClaimModifyVolumeInfeasible) {
+			pvcKey, err := cache.MetaNamespaceKeyFunc(pvc)
+			if err != nil {
+				return err
+			}
+			ctrl.uncertainPVCs[pvcKey] = *pvc.DeepCopy()
+		}
+	}
+
+	return nil
+}
+
+func (ctrl *modifyController) addPVC(obj interface{}) {
+	objKey, err := util.GetObjectKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.claimQueue.Add(objKey)
+}
+
+func (ctrl *modifyController) updatePVC(oldObj, newObj interface{}) {
+	oldPVC, ok := oldObj.(*v1.PersistentVolumeClaim)
+	if !ok || oldPVC == nil {
+		return
+	}
+
+	newPVC, ok := newObj.(*v1.PersistentVolumeClaim)
+	if !ok || newPVC == nil {
+		return
+	}
+
+	// Only trigger modify volume if the following conditions are met
+	// 1. Non empty vac name
+	// 2. oldVacName != newVacName
+	// 3. PVC is in Bound state
+	oldVacName := oldPVC.Spec.VolumeAttributesClassName
+	newVacName := newPVC.Spec.VolumeAttributesClassName
+	if newVacName != nil && *newVacName != "" && *newVacName != *oldVacName && oldPVC.Status.Phase == v1.ClaimBound {
```

Comment:
The `oldVacName` can be nil for existing PVC, right? In which case this code would crash?

Reply:
Does not hurt to add a case but can it though? The manual testing step I did, even the pvc does not have vac first, and then add it, the vac got updated successfully and there is no error.

Reply:
I am not sure if same flow would have been executed in manual testing. It is always best to guard against code that can cause outright crash.